fix: Abort webauthn prompt when switching factors

magizh-okta opened the pull request:
* In some browsers the webauthn prompt is subtle and users often go back to choose another factor.
When users do this the browser prompt is still present and it is an UX issue. Also when they come back to use webauthn again it errors out.

* This fix is to kill the webauthn prompt when switching factors.

* introduced a new error to prevent displaying an error message when error is due to abort via code since factor UI was switched.

**Enroll and verify with switching works fine** 
![enroll-abort](https://user-images.githubusercontent.com/17267130/66507698-889bb800-ea84-11e9-8887-2937d2d85d74.gif)
![verify-abort](https://user-images.githubusercontent.com/17267130/66507699-889bb800-ea84-11e9-8495-70f1260d9bf8.gif)

** verify chrome ** 
![verify-abort-cc](https://user-images.githubusercontent.com/17267130/66515162-6b221a80-ea93-11e9-8e44-a31f2e6cec9f.gif)

** enroll chrome (back is triggered via js in console) user cannot switch without cancel anyway ** 
![enrol-abort-cc](https://user-images.githubusercontent.com/17267130/66515151-665d6680-ea93-11e9-9a97-699bbd93cc36.gif)

**BEFORE** 
recording is for u2f but it is the same issue for webauthn https://okta.app.box.com/s/xg9kp9uaw2dsjks4n2manoe0iz2spf3f

resolves: OKTA-249465

magizh-okta commented on src/EnrollWebauthnController.js:
```diff
@@ -88,9 +93,14 @@ function (Okta, Errors, FormType, FormController, CryptoUtil, webauthn, Footer,
-                  xhr: {responseJSON: {errorSummary: error.message}}
-                });
+                // Do not display if it is abort error triggered by code when switching.
+                if (error && error.code === 20) {
```
@gowthamidommety-okta @vincentvoong-okta  we can actually determine what the error message is in each case. I got it wrong previously as the console statements printed an object differently. FF as a string and chrome as object.

magizh-okta replied:
Oops taking this back! Firefox triggers a different code vs chrome :/ /cc @mauriciocastillosilva-okta  trying to fix this now

mauriciocastillosilva-okta replied:
ah ok. Thanks for the update. Let me know what you find out, I'm curious about this

magizh-okta replied:
in chrome the error code was 20 for programatic abort and 0 for user abort. But in firefox it is always 20. ( which is per spec) so I don't want to rely on this in consistent behavior.

haishengwu-okta commented on src/EnrollWebauthnController.js:
```diff
@@ -79,7 +79,12 @@ function (Okta, Errors, FormType, FormController, CryptoUtil, webauthn, Footer,
               excludeCredentials: getExcludeCredentials(activation.excludeCredentials)
             });
-            return new Q(navigator.credentials.create({publicKey: options}))
+            var webauthnAbortController = new AbortController();
```
`AbortController` seems not compatible in older browser. would there be any regression concern?
https://www.caniuse.com/#feat=abortcontroller

magizh-okta replied:
@haishengwu-okta  older meaning IE 11 ? webauthn  itself wont work so we dont have to worry about it I guess. ie) we will safely fail before we use the abortController.

mauriciocastillosilva-okta commented on src/EnrollWebauthnController.js:
```diff
@@ -79,7 +79,12 @@ function (Okta, Errors, FormType, FormController, CryptoUtil, webauthn, Footer,
             });
-            return new Q(navigator.credentials.create({publicKey: options}))
+            var webauthnAbortController = new AbortController();
+            this.appState.set('webauthnAbortController', webauthnAbortController);
```
do this really belong to the `appState`? Can we instead abort in the Footer when the user clicks "Back to Factors"? It seems to me this is very specific to this factor

mauriciocastillosilva-okta commented on src/EnrollWebauthnController.js:
```diff
@@ -88,9 +93,14 @@ function (Okta, Errors, FormType, FormController, CryptoUtil, webauthn, Footer,
-                });
+                // Do not display if it is abort error triggered by code when switching.
+                if (error && error.code === 20) {
+                  throw new Errors.WebauthnAbortError();
```
its not really an error, is it? Can we just not do anything here? Or in other words, just handle this error when the code is not 20?

magizh-okta replied:
Since we use BaseLoginModel here there is a fail case in base login model and it expects some know error else it fails. 

we already have similar cases 
https://github.com/okta/okta-signin-widget/blob/master/src/models/BaseLoginModel.js#L37

mauriciocastillosilva-okta commented on src/EnrollChoicesController.js:
```diff
@@ -87,6 +87,10 @@ function (Okta, FormController, Enums, RouterUtil, FactorList,
       },
 
       initialize: function (options) {
+        if (options.appState.get('webauthnAbortController')) {
```
I think this should be handled in a similar way to how we cancel polling when we switch from Okta Verify to another factor. There is a `factorSwitched` event that is fired in [FactorDropdownOptions#L31](https://github.com/okta/okta-signin-widget/blob/master/src/views/mfa-verify/dropdown/FactorsDropDownOptions.js#L31) that we listen to in [Factor#315](https://github.com/okta/okta-signin-widget/blob/master/src/models/Factor.js#L315), so we can stop polling. Sounds like we can do something similar?

magizh-okta replied:
nice! I reused this and moved away from appstate as well.

haishengwu-okta replied:
curious why you cannot extend `src/models/Factor.js` as Mau mentioned but have to listen to a new event when click back?

magizh-okta replied:
@haishengwu-okta  I am reusing the event in verify flow. But during the enroll when back is clicked I don't see any event being triggered.

haishengwu-okta replied:
I meant `backToFactors`. looks like enroll and challenge flow need different event?

mauriciocastillosilva-okta replied:
I think this is right, since those 2 events are different:
- `factorSwitched` is the event we have when users select a new factor in the beacon dropdown. This is only in verify flows
- `backToFactors` is the new event created in this PR, this is only in enroll flows (there is no beacon dropdown in enroll flows)

magizh-okta replied:
yeah in verify we switchFactors and enroll we can go back to factorlist.

mauriciocastillosilva-okta commented on src/views/mfa-verify/dropdown/FactorsDropDownOptions.js:
```diff
@@ -27,7 +27,7 @@ define(['okta', 'util/RouterUtil'], function (Okta, RouterUtil) {
     var url = RouterUtil.createVerifyUrl(model.get('provider'), factorType, factorIndex);
     var self = this;
-
+    
```
please revert this change

mauriciocastillosilva-okta commented on test/unit/spec/EnrollWebauthn_spec.js:
```diff
@@ -88,17 +88,19 @@ function (Okta,
-            clientDataJSON: CryptoUtil.strToBin(testClientData)
-          }
-        });
+        if (!doNotResolve) {
```
2 comments here:
1. can we make this not a "double negative" -> if NOT doNotResolve... maybe the arg can be named something like `resolvePromise` or `successfullyResolvePromise`?
2. If it does not resolve, then we are timing out? Not sure how that is handled in the test...

magizh-okta replied:
we will be in waiting state. More like faking the state where we see the webauthn prompt but user has not yet authenticated via touchid/u2f key

mauriciocastillosilva-okta commented on test/unit/spec/VerifyWebauthn_spec.js:
```diff
@@ -14,7 +14,7 @@ define([
   'util/webauthn',
   'helpers/xhr/MFA_REQUIRED_allFactors',
-  'helpers/xhr/MFA_REQUIRED_multipleFactors',
+  'helpers/xhr/MFA_REQUIRED_multipleWebauthn_question',
```
can we also rename the variable we use? It's just easier in the future to know what the mock response is if the name matches the actual mock response

mauriciocastillosilva-okta commented on test/unit/spec/VerifyWebauthn_spec.js:
```diff
@@ -53,6 +53,14 @@ function (Okta,
+  function clickFactorInDropdown (test, factorName) {
+    if (factorName === 'WEBAUTHN') {
+      test.beacon.getOptionsLinks().eq(0).click();
+    } else {
```
nit: check "QUESTION", so we know it is that factor we are selecting

magizh-okta commented:
@mauriciocastillosilva-okta  I addressed comments and added some cleanup code + handled signOut case too. 

![singout](https://user-images.githubusercontent.com/17267130/66599296-bdc50a80-eb57-11e9-829e-e0b482752fe3.gif)

mauriciocastillosilva-okta commented on test/unit/spec/EnrollWebauthn_spec.js:
```diff
@@ -174,13 +176,34 @@ function (Okta,
         });
       });
+      
+      itp('calls aborts on appstate when switching to factor list after clicking enroll', function () {
```
singular abort: `calls abort on appsate`...
